Add support for copyTexSubImage2D, misc fixes

1chandu opened the pull request:
<!-- For feature, feature enhancement or bug fix, create an issue first and finish To Do List there -->
<!-- Anything doesn't work as expected is a bug, including code, doc and test -->
For #800
<!-- For other PRs without open issue -->
#### Background
Fixes for issues found while working for #800

1. Add support for `copyTexSubImage2D` for Texture class.
2. Fix texture binding for `copyTexImage2D` (pass binding index for bind call)
3. Remove un-used `offset` argument form `copyTexImage2D`
4. Fix Framebuffer's `_unattach` method
5. Framebuffer: make attachment re-size optional during attach() , default behavior un-changed.

<!-- For all the PRs -->
#### Change List
- Add support for copyTexSubImage2D, misc fixes

ibgreen commented on docs/api-reference/webgl/texture.md:
```diff
@@ -181,6 +180,31 @@ The following WebGL APIs are called in the function
 
+### copySubFramebuffer
+
+Updates defined area of a two-dimensional texture image or cube-map texture image with pixels from the current framebuffer (rather than from client memory). (gl.copyTexSubImage2D wrapper)
```
Nits:
* "Updates defined area" => "Copies pixels from the current framebuffer into the specified area of a ..."
* "current framebuffer" => "this framebuffer"?

ibgreen commented on docs/api-reference/webgl/texture.md:
```diff
@@ -181,6 +180,31 @@ The following WebGL APIs are called in the function
 [gl.copyTexImage2D](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/copyTexImage2D), [gl.bindFramebuffer](https://developer.mozilla.org/en-US/docs/Web/API/WebGLRenderingContext/bindFramebuffer)
 
 
+### copySubFramebuffer
```
Nit: Add a signature `(opts: Object)`?

ibgreen commented on modules/core/src/webgl/framebuffer.js:
```diff
@@ -668,8 +669,17 @@ export default class Framebuffer extends Resource {
   }
 
   _unattach({attachment}) {
```
Nit/Optimization opportunity: Since private method with few params, no need to incur overhead of named parameters.

1chandu commented:
@ibgreen addressed your comments.

> Can't the copySubFramebuffer function simply be implemented using a call to framebuffer.copyToTexture?

Good catch !! Interesting, `Framebuffer` is implementing `copyTexSubImage2D` and `Texture` is implementing `copyTexImage2D`. We should move this to same class. Also will be resolved if we make all these wrappers into global functions.

> Would it be possible to unify copyFramebuffer and copySubFramebuffer into a single method,

We don't have to do this given I am now using Framebuffer.copyToTexture, but for discussion purpose, I think  we can't, given copyFramebuffer allocates texture storage and copySubFramebuffer doesn't. You might have a use case where you will be copying entire image but don't want to allcoate memory, can't distinguish this case based on existing parameters.

ibgreen commented:
> e, I think we can't, given copyFramebuffer allocates texture storage and copySubFramebuffer doesn't. You might have a use case where you will be copying entire image but don't want to allcoate memory, can't distinguish this case based on existing parameters.

Hmm, I must admit that I am not fresh on the details, but those sound like subtle differences that I think will confuse most users, whether we expose one or two functions.

My sense is that this could make it even more valuable to unify the two functions, we can provide an extra parameter to control the allocation and describe it in the docs. 

Since we are heading for 7.0 this is a good time to do an API refinement. Let's sync off-line.

1chandu commented:
Remaining issue tracked here : https://github.com/uber/luma.gl/issues/806
